vm/actor/src/builtin/miner/deadline_assignment.rs

```diff
@@ -0,0 +1,185 @@
+            // compaction as long as possible.
+            a.partitions_after_assignment(partition_size)
+                .cmp(&b.partitions_after_assignment(partition_size))
+        })
+        .then_with(|| {
+            // Ok, we'll end up with the same number of partitions any which way we
+            // go. Try to fill up a partition instead of opening a new one.
+            a.is_full_now(partition_size)
+                .cmp(&b.is_full_now(partition_size))
+        })
+        .then_with(|| {
+            // Either we have two open partitions, or neither deadline has an open
+            // partition.
+
+            // If we have two open partitions, fill the deadline with the most-full
+            // open partition. This helps us assign runs of sequential sectors into
+            // the same partition.
+            if !a.is_full_now(partition_size) && !b.is_full_now(partition_size) {
+                a.total_sectors.cmp(&b.total_sectors).reverse()
+            } else {
+                Ordering::Equal
+            }
+        })
+        .then_with(|| {
+            // Otherwise, assign to the deadline with the least live sectors. This
+            // will break the tie in one of the two immediately preceding
+            // conditions.
+            a.live_sectors.cmp(&b.live_sectors)
+        })
+        .then_with(|| {
+            // Finally, fall back on the deadline index.
+            a.index.cmp(&b.index)
+        })
+}
+
+// Assigns partitions to deadlines, first filling partial partitions, then
+// adding new partitions to deadlines with the fewest live sectors.
+pub fn assign_deadlines(
+    partition_size: u64,
+    deadlines: &[Option<Deadline>],
+    sectors: Vec<SectorOnChainInfo>,
+) -> Vec<Vec<SectorOnChainInfo>> {
+    struct Entry {
+        partition_size: u64,
+        info: DeadlineAssignmentInfo,
+    }
+
+    impl PartialEq for Entry {
+        fn eq(&self, other: &Self) -> bool {
+            self.cmp(other) == Ordering::Equal
+        }
+    }
+
+    impl Eq for Entry {}
+
+    impl PartialOrd for Entry {
+        fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+            Some(self.cmp(other))
+        }
+    }
+
+    impl Ord for Entry {
+        fn cmp(&self, other: &Self) -> Ordering {
+            // we're using a max heap instead of a min heap, so we need to reverse the ordering
+            cmp(&self.info, &other.info, self.partition_size).reverse()
+        }
+    }
+
+    let mut heap: BinaryHeap<Entry> = deadlines
+        .iter()
+        .enumerate()
+        .filter_map(|(index, deadline)| deadline.as_ref().map(|dl| (index, dl)))
+        .map(|(index, deadline)| Entry {
+            partition_size,
+            info: DeadlineAssignmentInfo {
+                index,
+                live_sectors: deadline.live_sectors,
+                total_sectors: deadline.total_sectors,
+            },
+        })
+        .collect();
+
+    assert!(!heap.is_empty());
```

Comment:
why is this assertion okay? Can you be sure that deadlines will always not be empty? I don't see how this is guaranteed, but maybe I'll find the answer as I review more

Reply:
I dont think the assertion is really necessary, but it is true that the deadlines array is none empty. See miner_state.rs:422

Reply:
Yeah, I know that, but I wasn't sure if the array of `None` deadlines translated into the deadline_array variable that is passed into the function. I'm checking that file now though

Reply:
This definitely needs a comment, let me look into why I put this here

Reply:
So as far as I can see, the Go impl implicitly asserts that the heap is non-empty by accessing the top element at https://github.com/filecoin-project/specs-actors/blob/master/actors/builtin/miner/deadline_assignment.go#L204

Reply:
Which is more or less equivalent to our `heap.peek_mut().unwrap()`. So afaics it really doesn't matter whether that assert is there or not, either way it will be a runtime error if the heap is empty.